src/libstd/sys/sgx/rwlock.rs

```diff
@@ -239,7 +241,7 @@ mod tests {
             zero_stack();
             let mut init = MaybeUninit::<RWLock>::zeroed();
             rwlock_new(&mut init);
-            assert_eq!(mem::transmute::<_, [u8; 128]>(init.assume_init()).as_slice(), RWLOCK_INIT)
+            assert_eq!(mem::transmute::<_, [u8; 144]>(init.assume_init()).as_slice(), RWLOCK_INIT)
```

Comment:
I don't really understand this test, and it's "zero the stack" thing - if some bytes are uninitialized, the comparison is UB, and there's no real way around that (without the proposed LLVM `freeze` operation), AFAIK.

And `MaybeUninit::<RWLock>::zeroed()` has the same semantics as `[0u8; 144]`, doesn't it? So there's no way any bytes wouldn't be zeroed.

I know this is offtopic, but I'm a bit worried this test (and perhaps more like it?) isn't actually testing anything (cc @RalfJung)

Reply:
> I don't really understand this test, and it's "zero the stack" thing

`transmute::<_, RWLock>(*RWLOCK_INIT)` needs to be equivalent to `RWLock::new()`. This test tries to assert that. Since part of the representation of `RWLock::new()` may contain uninitialized memory, some trickery (and UB) is needed, but this (trying very hard to get the uninitialized memory to be 0) is the best we can do.

> if some bytes are uninitialized, the comparison is UB, and there's no real way around that (without the proposed LLVM freeze operation), AFAIK.

Yes, this is exactly what the comment just above this code block tries to explain (“issue with the test code”). However, this test has been working fine and this is the best we can do. If changes to codegen are made that make this test fail, we can consider other options at that time. Note that UB *only* occurs during the execution of this test, normal use of this primitive never uses uninitialized memory.

> And MaybeUninit::<RWLock>::zeroed() has the same semantics as [0u8; 144], doesn't it? So there's no way any bytes wouldn't be zeroed.

I'm not sure what you mean by this. Where are we using `[0u8; 144]`?

Reply:
> And MaybeUninit::<RWLock>::zeroed() has the same semantics as [0u8; 144], doesn't it?

It's not; the former resets any padding to `Undef` as happens any time a struct is copied.

Reply:
> It's not; the former resets any padding to `Undef` as happens any time a struct is copied.

Oh right I forgot about that (sadly we don't tell LLVM about this, since we just use a plain `memcpy`).

Wouldn't that also imply that it's impossible for `RWLock::new()` to guarantee the padding bytes of `RWLock` are initialized to 0, which this tests tries to prove (and arguably fails because its comparison of undef bytes is UB)?

Reply:
> Wouldn't that also imply that it's impossible for RWLock::new() to guarantee the padding bytes of RWLock are initialized to 0

Yes. (I am not sure if `RwLock` has any padding bytes, though.)

Reply:
If we want to test for "no padding bytes", I wonder if this test should rely on const-eval instead, e.g. transmuting to `[u8; 144]` in a `const`, in a way that would cause an error if any of the bytes were `undef`.

Reply:
@jethrogb is it crucial that `RwLock::new()` returns something all-0, or is it just crucial that using something all-0 instead of `RwLock::new()` must be fine? If the answer is the latter, when padding wouldn't actually be a problem (except for the test).

Reply:
Okay turns out I was very misled by the "try hard to make sure" comment - it doesn't describe what is being tested, but specifically the `zero_stack()` call.

What this test appears to be checking is that the hardcoded bytes in `RWLOCK_INIT` (which presumably are also hardcoded on the C side) are the same as `RWLock::new()` *ignoring* padding.

But there's no way to automatically ignore padding, so what's being done here is trying to zero the stack space where `RWLock::new()` (and all of its transitive callers) writes every part of the `RWLock`, so that the padding likely stays zeroed.

I'm less worried about the test, but the comment could probably be improved to make it clearer that `RWLock::new()` having padding left uninitialized is not an issue, it just makes testing the initialized bytes harder.

I think the simplest way to get zeroed data (although I don't know how UB it is) is to use a `static`, e.g.: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3abaa186579b5522baad9f492a784087 (that works even on stable, heh).

Reply:
> is it crucial that RwLock::new() returns something all-0, or is it just crucial that using something all-0 instead of RwLock::new() must be fine?

The latter

> the comment could probably be improved

@eddyb I think you and @RalfJung have a pretty good idea of what I'm trying to achieve here, and you also have a better understanding of undef than I have, would you mind drafting a comment that makes sense to you?

> I think the simplest way to get zeroed data (although I don't know how UB it is) is to use a `static`

Neat, I think this is a pretty good way.

Reply:
> would you mind drafting a comment that makes sense to you?

Well, if the `static` works, it's easier to explain that the `undef` parts will get zero-initialized *in practice*.

To be honest, one of the most confusing things to me was that `zero_stack();` didn't have an empty line after it, which made me assume incorrectly what the comment referred to (if I remove the `zero_stack()` call, the test fails, and I think when I saw that I understood what the comment meant).

A `static` should be clearer from the start compared to zeroing the stack in an unguaranteed way.